vllm/model_executor/layers/quantization/moe_quant_int.py

```diff
@@ -0,0 +1,362 @@
+        linear_quant_method = cls.get_from_keys(config, ["quant_method"])
+        weight_bits = cls.get_from_keys(config, ["bits"])
+        group_size = cls.get_from_keys(config, ["group_size"])
+        lm_head_quantized = cls.get_from_keys_or(config, ["lm_head"],
+                                                 default=False)
+        if linear_quant_method == "gptq":
+            has_zp = not cls.get_from_keys(config, ["sym"])
+            modules_to_not_convert = []
+        elif linear_quant_method == "awq":
+            has_zp = cls.get_from_keys(config, ["zero_point"])
+            modules_to_not_convert = cls.get_from_keys(
+                config, ["modules_to_not_convert"])
+        else:
+            raise ValueError("moe_quant_int only support gptq and awq.")
+
+        return cls(linear_quant_method, weight_bits, group_size, has_zp,
+                   lm_head_quantized, modules_to_not_convert, config)
+
+    @classmethod
+    def override_quantization_method(cls, hf_quant_cfg,
+                                     user_quant) -> Optional[str]:
+        can_convert = cls.is_moe_quant_int_compatible(hf_quant_cfg)
+        is_valid_user_quant = (user_quant is None
+                               or user_quant == "moe_quant_int")
+        if can_convert and is_valid_user_quant:
+            return cls.get_name()
+        return None
+
+    @classmethod
+    def is_moe_quant_int_compatible(cls, quant_config: Dict[str, Any]):
+        # Extract data from quant config.
+        quant_method = quant_config.get("quant_method", "").lower()
+        num_bits = quant_config.get("bits")
+        desc_act = quant_config.get("desc_act")
+
+        gptq_compatible = quant_method == "gptq" and \
+                not desc_act and num_bits in [4, 8] and \
+                GPTQMarlinConfig.is_gptq_marlin_compatible(quant_config)
+        awq_compatible = quant_method == "awq" and \
+                num_bits == 4 and \
+                AWQMarlinConfig.is_awq_marlin_compatible(quant_config)
+
+        return gptq_compatible or awq_compatible
+
+    def get_quant_method(self, layer: torch.nn.Module,
+                         prefix: str) -> Optional["QuantizeMethodBase"]:
+        if is_layer_skipped_quant(prefix, self.modules_to_not_convert):
+            return UnquantizedLinearMethod()
+        elif isinstance(layer, LinearBase):
+            if self.linear_quant_method == "gptq":
+                gptq_config = GPTQMarlinConfig.from_config(self.full_config)
+                return GPTQMarlinLinearMethod(gptq_config)
+            elif self.linear_quant_method == "awq":
+                awq_config = AWQMarlinConfig.from_config(self.full_config)
+                return AWQMarlinLinearMethod(awq_config)
+            else:
+                raise ValueError("moe_quant_int only support gptq and awq.")
+        elif isinstance(layer, FusedMoE):
+            return MoeQuantIntMethod(self)
+        return None
```

Comment:
This is an interesting hack - I wonder if we could just enable the `MoeQuantIntMethod` as a condition inside of the other quantization methods rather than duplicating them here in this config

Reply:
I considered this before, but I created a new quantization method finally. The reasons are

1. This quantization method can be combined all gptq/awq quanzation methods, should we add it to all quanzaition methods that supported gptq/awq, or just gptq-marlin/awq-marlin ?
2. This quantization method and triton kernel use a different weight format, it is just compatible with gptq/awq, and accept gpt/awq weight.
3. Make the code more clear and easy to maintenance (less duplication)